scapy/contrib/ikev2.py

```diff
@@ -399,62 +428,63 @@
     142: "Robust Header Compression",
 }
 
-# the name 'IKEv2TransformTypes' is actually a misnomer (since the table
-# holds info for all IKEv2 Attribute types, not just transforms, but we'll
-# keep it for backwards compatibility... for now at least
-IKEv2TransformTypes = IKEv2AttributeTypes
-
-IKEv2TransformNum = {}
-for n in IKEv2TransformTypes:
-    val = IKEv2TransformTypes[n]
-    tmp = {}
-    for e in val[1]:
-        tmp[val[1][e]] = e
-    IKEv2TransformNum[val[0]] = tmp
-
-IKEv2Transforms = {}
-for n in IKEv2TransformTypes:
-    IKEv2Transforms[IKEv2TransformTypes[n][0]] = n
-
-del n
-del e
-del tmp
-del val
-
-# Note: Transform and Proposal can only be used inside the SA payload
-IKEv2_payload_type = ["None", "", "Proposal", "Transform"]
+IKEv2PayloadTypes = {
+    0: "None",
+    2: "Proposal",   # used only inside the SA payload
+    3: "Transform",  # used only inside the SA payload
+    33: "SA",
+    34: "KE",
+    35: "IDi",
+    36: "IDr",
+    37: "CERT",
+    38: "CERTREQ",
+    39: "AUTH",
+    40: "Nonce",
+    41: "Notify",
+    42: "Delete",
+    43: "VendorID",
+    44: "TSi",
+    45: "TSr",
+    46: "Encrypted",
+    47: "CP",
+    48: "EAP",
+    49: "GSPM",
+    50: "IDg",
+    51: "GSA",
+    52: "KD",
+    53: "Encrypted_Fragment",
+    54: "PS"
+}
 
-IKEv2_payload_type.extend([""] * 29)
-IKEv2_payload_type.extend(["SA", "KE", "IDi", "IDr", "CERT", "CERTREQ", "AUTH", "Nonce", "Notify", "Delete",  # noqa: E501
-                           "VendorID", "TSi", "TSr", "Encrypted", "CP", "EAP", "", "", "", "", "Encrypted_Fragment"])  # noqa: E501
 
-IKEv2_exchange_type = [""] * 34
-IKEv2_exchange_type.extend(["IKE_SA_INIT", "IKE_AUTH", "CREATE_CHILD_SA",
-                            "INFORMATIONAL", "IKE_SESSION_RESUME"])
+IKEv2ExchangeTypes = {
+    34: "IKE_SA_INIT",
+    35: "IKE_AUTH",
+    36: "CREATE_CHILD_SA",
+    37: "INFORMATIONAL",
+    38: "IKE_SESSION_RESUME",
+    43: "IKE_INTERMEDIATE"
+}
 
 
 class IKEv2_class(Packet):
     def guess_payload_class(self, payload):
         np = self.next_payload
-        logging.debug("For IKEv2_class np=%d", np)
         if np == 0:
             return conf.raw_layer
-        elif np < len(IKEv2_payload_type):
-            pt = IKEv2_payload_type[np]
-            logging.debug(globals().get("IKEv2_payload_%s" % pt, IKEv2_payload))  # noqa: E501
-            return globals().get("IKEv2_payload_%s" % pt, IKEv2_payload)
         else:
-            return IKEv2_payload
+            pt = IKEv2PayloadTypes.get(np, "Payload")
+            return globals().get("IKEv2_" + pt, IKEv2_Payload)
```

Comment:
I would really prefer changing this to a `bind_layers` way of doing things, like my ISAKMP PR, than keeping it. It's not scapy-like and kind of crappy.

Reply:
In the second part I refactored the code to use `bind_layers`. If you want to see the full picture, please take a look at #3801, in particular commit [937afef5bd - IKEv2: bind all layers together correctly](https://github.com/secdev/scapy/pull/3801/commits/937afef5bdfa24b72d388630446abe921fe503d0)). 

Maybe I really should join the second pull request with the first one? Initially, I thought it migth be better to merge little parts, but it seems to be more confusing than helpful, what do you think?

Anyway, this is how the code currrently looks like in #3801: 

https://github.com/secdev/scapy/blob/a66ce8b213b7853b8385bc0a2726ce270ebb4228/scapy/contrib/ikev2.py#L868-L872

If you prefer that I convert the loop in a manually written list, that would be ok for me, too.

Reply:
_(the loop has the advantage that it is less redundand, because the binding information already exists in the `IKEv2PayloadTypes` map)_

Reply:
If you can merge everything that'll help. Thanks

Reply:
And yeah, I prefer a manual list rather than a loop.

Reply:
Ok, thanks! I'll join the other PR with this one and close it. And I'll rework the binding.

Reply:
Done. 

_Fun fact: I created the 'manual' bindings from the loop by converting the bind_layers statement into a print statement and copying the output:_

```python
# bind all IKEv2 payload classes together
for np, name in IKEv2PayloadTypes.items():
    cls = globals().get("IKEv2_" + name)
    if cls:
        print(f"bind_layers(_IKEv2_Packet, IKEv2_{name}, next_payload={np})")
   ```